File: pylox/pylox/Parser.py
from typing import Callable, Final, Optional

from Expr import (
    Assign,
    Binary,
    Call,
    Expr,
    Get,
    Grouping,
    Literal,
    Logical,
    Set,
    Super,
    This,
    Unary,
    Variable,
)
from Token import Token
from TokenTypes import TokenType
from Stmt import Block, Class, Expression, Function, If, Return, Stmt, Print, Var, While


class Parser:
    class ParseError(RuntimeError): ...

    def __init__(
        self,
        tokens: list[Token],
        parse_error: Callable[[Token, str], None],
    ) -> None:
        self._tokens: Final[list[Token]] = tokens
        self._current: int = 0
        self._parse_error: Final[Callable[[Token, str], None]] = parse_error

    def parse(self) -> list[Stmt]:
        statements: list[Stmt] = []

        while not self._is_at_end():
            statements.append(self._declaration())  # type: ignore[reportArgumentType]

        return statements

    def _expression(self):
        return self._assignment()

    def _assignment(self) -> Expr:
        expr: Expr = self._or()

        if self._match(TokenType.EQUAL):
            equals: Token = self._previous()
            value: Expr = self._assignment()

            if isinstance(expr, Variable):
                return Assign(expr.name, value)
            elif isinstance(expr, Get):
                return Set(expr.object, expr.name, value)

            self._error(equals, "Invalid assignment target.")

        return expr

    def _or(self) -> Expr:
        expr: Expr = self._and()

        while self._match(TokenType.OR):
            operator = self._previous()
            right = self._and()
            expr = Logical(expr, operator, right)

        return expr

    def _and(self) -> Expr:
        expr: Expr = self._equality()

        while self._match(TokenType.AND):
            operator = self._previous()
            right = self._equality()
            expr = Logical(expr, operator, right)

        return expr

    def _declaration(self):
        try:
            if self._match(TokenType.CLASS):
                return self._class_declaration()
            if self._match(TokenType.FUN):
                return self._function("function")
            if self._match(TokenType.VAR):
                return self._var_declaration()
            return self._statement()
        except self.ParseError:
            self._synchronize()
            return None

    def _class_declaration(self) -> Class:
        name = self._consume(TokenType.IDENTIFIER, "Expect class name.")
        super_class: Optional[Variable] = None
        if self._match(TokenType.LESS):
            self._consume(TokenType.IDENTIFIER, "Expect superclass name.")
            super_class = Variable(self._previous())
        self._consume(TokenType.LEFT_BRACE, "Expect '{' before class body.")
        methods: list[Function] = []
        while not self._check(TokenType.RIGHT_BRACE) and not self._is_at_end():
            methods.append(self._function("method"))

        self._consume(TokenType.RIGHT_BRACE, "Expect '}' after class body.")
        return Class(name, super_class, methods)

    def _function(self, kind: str) -> Function:
        name: Token = self._consume(TokenType.IDENTIFIER, f"Expect {kind} name.")
        self._consume(TokenType.LEFT_PAREN, f"Expect '(' after {kind} name.")

        arguments: list[Token] = []
        if not self._check(TokenType.RIGHT_PAREN):
            arguments.append(
                self._consume(TokenType.IDENTIFIER, "Expect parameter name.")
            )
            while self._match(TokenType.COMMA):
                if len(arguments) >= 255:
                    self._error(self._peek(), "Can't have more than 255 parameters.")
                arguments.append(
                    self._consume(TokenType.IDENTIFIER, "Expect parameter name.")
                )

        self._consume(TokenType.RIGHT_PAREN, "Expect ')' after parameters.")
        self._consume(TokenType.LEFT_BRACE, f"Expect '{{' before {kind} body.")
        body: list[Stmt] = self._block()
        return Function(name, arguments, body)

    def _statement(self) -> Stmt:
        if self._match(TokenType.FOR):
            return self._for_statement()
        if self._match(TokenType.RETURN):
            return self._return_statement()
        if self._match(TokenType.WHILE):
            return self._while_statement()
        if self._match(TokenType.IF):
            return self._if_statement()
        if self._match(TokenType.PRINT):
            return self._print_statement()
        if self._match(TokenType.LEFT_BRACE):
            return Block(self._block())

        return self._expression_statement()

    def _for_statement(self) -> Stmt:
        self._consume(TokenType.LEFT_PAREN, "Expect '(' after 'for'.")

        initializer: Optional[Stmt]
        if self._match(TokenType.SEMICOLON):
            initializer = None
        elif self._match(TokenType.VAR):
            initializer = self._var_declaration()
        else:
            initializer = self._expression_statement()

        condition: Optional[Expr] = None
        if not self._check(TokenType.SEMICOLON):
            condition = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after loop condition.")

        increment: Optional[Expr] = None
        if not self._check(TokenType.RIGHT_PAREN):
            increment = self._expression()
        self._consume(TokenType.RIGHT_PAREN, "Expect ')' after for clauses.")

        body: Stmt = self._statement()

        if increment:
            body = Block([body, Expression(increment)])

        if not condition:
            condition = Literal(True)

        body = While(condition, body)

        if initializer:
            body = Block([initializer, body])

        return body

    def _if_statement(self) -> Stmt:
        self._consume(TokenType.LEFT_PAREN, "Expect '(' after 'if'.")
        condition = self._expression()
        self._consume(TokenType.RIGHT_PAREN, "Expect ')' after if condition.")

        then_branch: Stmt = self._statement()
        else_branch: Stmt = None  # type: ignore[reportAssignmentType]
        if self._match(TokenType.ELSE):
            else_branch = self._statement()

        return If(condition, then_branch, else_branch)

    def _print_statement(self):
        value: Expr = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after value.")
        return Print(value)

    def _return_statement(self) -> Stmt:
        keyword: Token = self._previous()
        value: Expr = None  # type: ignore[reportAssignmentType]
        if not self._check(TokenType.SEMICOLON):
            value = self._expression()

        self._consume(TokenType.SEMICOLON, "Expect ';' after return value.")
        return Return(keyword, value)

    def _var_declaration(self):
        name: Token = self._consume(TokenType.IDENTIFIER, "Expect variable name.")

        intializer: Expr = None  # type: ignore[reportAssignmentType]
        if self._match(TokenType.EQUAL):
            intializer = self._expression()

        self._consume(TokenType.SEMICOLON, "Expect ';' after variable declaration.")
        return Var(name, intializer)

    def _while_statement(self):
        self._consume(TokenType.LEFT_PAREN, "Expect '(' after 'while'.")
        condition = self._expression()
        self._consume(TokenType.RIGHT_PAREN, "Expect ')' after condition.")
        body = self._statement()

        return While(condition, body)

    def _expression_statement(self):
        value: Expr = self._expression()
        self._consume(TokenType.SEMICOLON, "Expect ';' after expression.")
        return Expression(value)

    def _block(self):
        statements: list[Stmt] = []

        while not self._check(TokenType.RIGHT_BRACE) and not self._is_at_end():
            statements.append(self._declaration())  # type: ignore[reportArgumentType]

        self._consume(TokenType.RIGHT_BRACE, "Expect '}' after block.")
        return statements

    def _equality(self) -> Expr:
        # equality       → comparison ( ( "!=" | "==" ) comparison )* ;
        expr: Expr = self._comparision()

        while self._match(TokenType.EQUAL_EQUAL, TokenType.BANG_EQUAL):
            operator: Token = self._previous()
            right: Expr = self._comparision()
            expr = Binary(expr, operator, right)

        return expr

    def _comparision(self) -> Expr:
        # comparison     → term ( ( ">" | ">=" | "<" | "<=" ) term )* ;
        expr: Expr = self._term()

        while self._match(
            TokenType.GREATER,
            TokenType.GREATER_EQUAL,
            TokenType.LESS,
            TokenType.LESS_EQUAL,
        ):
            operator: Token = self._previous()
            right: Expr = self._term()
            expr = Binary(expr, operator, right)

        return expr

    def _term(self) -> Expr:
        # term           → factor ( ( "-" | "+" ) factor )* ;
        expr: Expr = self._factor()
        while self._match(
            TokenType.MINUS,
            TokenType.PLUS,
        ):
            operator: Token = self._previous()
            right: Expr = self._factor()
            expr = Binary(expr, operator, right)

        return expr

    def _factor(self) -> Expr:
        # factor         → unary ( ( "/" | "*" ) unary )* ;
        expr: Expr = self._unary()

        while self._match(
            TokenType.SLASH,
            TokenType.STAR,
        ):
            operator: Token = self._previous()
            right: Expr = self._unary()
            expr = Binary(expr, operator, right)

        return expr

    def _unary(self) -> Expr:
        # unary          → ( "!" | "-" ) unary
        #                  | primary ;
        if self._match(TokenType.BANG, TokenType.MINUS):
            operator: Token = self._previous()
            right: Expr = self._unary()
            return Unary(operator, right)

        return self._call()

    def _call(self) -> Expr:
        expr: Expr = self._primary()

        while True:
            if self._match(TokenType.LEFT_PAREN):
                expr = self._finish_call(expr)
            elif self._match(TokenType.DOT):
                name = self._consume(
                    TokenType.IDENTIFIER, "Expect property name after '.'."
                )
                expr = Get(expr, name)
            else:
                break

        return expr

    def _finish_call(self, expr: Expr) -> Expr:
        arguments: list[Expr] = []

        if not self._check(TokenType.RIGHT_PAREN):
            arguments.append(self._expression())

            while self._match(TokenType.COMMA):
                if len(arguments) >= 255:
                    self._error(self._peek(), "Can't have more than 255 arguments.")
                arguments.append(self._expression())

        paren: Token = self._consume(
            TokenType.RIGHT_PAREN, "Expect ')' after arguments."
        )

        return Call(expr, paren, arguments)

    def _primary(self) -> Expr:
        # primary        → NUMBER | STRING | "true" | "false" | "nil"
        #                  | "(" expression ")" ;
        if self._match(TokenType.TRUE):
            return Literal(True)
        if self._match(TokenType.FALSE):
            return Literal(False)
        if self._match(TokenType.NIL):
            return Literal(None)
        if self._match(TokenType.IDENTIFIER):
            return Variable(self._previous())
        if self._match(TokenType.NUMBER, TokenType.STRING):
            return Literal(self._previous().literal)
        if self._match(TokenType.THIS):
            return This(self._previous())

        if self._match(TokenType.LEFT_PAREN):
            expr: Expr = self._expression()
            self._consume(TokenType.RIGHT_PAREN, "Expect ')' after expression.")
            return Grouping(expr)

        if self._match(TokenType.SUPER):
            keyword = self._previous()
            self._consume(TokenType.DOT, "Expect '.' after 'super'.")
            method = self._consume(TokenType.IDENTIFIER, "Expect superclass method name.")
            return Super(keyword, method)

        raise self._error(self._peek(), "Expect expression.")

    def _consume(self, type: TokenType, message: str):
        if self._check(type):
            return self._advance()

        raise self._error(self._peek(), message)

    def _error(
        self,
        token: Token,
        message: str,
    ) -> ParseError:
        self._parse_error(token, message)
        return self.ParseError()

    def _synchronize(self) -> None:
        self._advance()

        while not self._is_at_end():
            if self._previous().type == TokenType.SEMICOLON:
                return

            match self._peek().type:
                case (
                    TokenType.CLASS
                    | TokenType.FUN
                    | TokenType.VAR
                    | TokenType.FOR
                    | TokenType.IF
                    | TokenType.WHILE
                    | TokenType.PRINT
                    | TokenType.RETURN
                ):
                    return
                case _:
                    self._advance()

    def _match(self, *types: TokenType) -> bool:
        for type in types:
            if self._check(type):
                self._advance()
                return True
        return False

    def _check(self, type: TokenType) -> bool:
        if self._is_at_end():
            return False
        return self._peek().type == type

    def _advance(self) -> Token:
        if not self._is_at_end():
            self._current += 1
        return self._previous()

    def _is_at_end(self) -> bool:
        return self._tokens[self._current].type == TokenType.EOF

    def _peek(self) -> Token:
        return self._tokens[self._current]

    def _previous(self) -> Token:
        return self._tokens[self._current - 1]

File: golox/internal/parser/parser.go
package parser

import (
	"fmt"
	"log"

	"github.com/englishlayup/crafting_interpreters/golox/internal/expr"
	"github.com/englishlayup/crafting_interpreters/golox/internal/stmt"
	"github.com/englishlayup/crafting_interpreters/golox/internal/token"
)

type ParseError struct {
}

func (p *ParseError) Error() string {
	return ""
}

type Parser struct {
	tokens      []token.Token
	current     int
	reportError func(token.Token, string)
}

func NewParser(tokens []token.Token, reportError func(token.Token, string)) Parser {
	return Parser{
		tokens:      tokens,
		reportError: reportError,
	}
}

func (p *Parser) Parse() []stmt.Stmt[any] {
	statements := make([]stmt.Stmt[any], 0)

	for p.isAtEnd() {
		statements = append(statements, p.declaration())
	}
	return statements
}

func (p *Parser) handleParseError(err error) {
    if _, ok := err.(*ParseError); ok {
        p.synchronize()
    }
    log.Fatal(err)
}

func (p *Parser) declaration() stmt.Stmt[any] {
	if p.match(token.CLASS) {
		decl, err := p.classDeclaration()
		if err != nil {
            p.handleParseError(err)
            return nil
		}
		return decl
	}
	if p.match(token.FUN) {
		decl, err := p.function("function")
		if err != nil {
            p.handleParseError(err)
            return nil
		}
		return decl
	}
	if p.match(token.VAR) {
		decl, err := p.varDelaration()
		if err != nil {
            p.handleParseError(err)
            return nil
		}
		return decl
	}
	return nil
}

func (p Parser) varDelaration() (stmt.Var[any], error) {
	name, err := p.consume(token.IDENTIFIER, "Expect variable name.")
	if err != nil {
		return stmt.Var[any]{}, err
	}
	var initializer expr.Expr[any]
	if p.match(token.EQUAL) {
		initializer, err = p.expression()
		if err != nil {
			return stmt.Var[any]{}, err
		}
	}
	if _, err := p.consume(
		token.SEMICOLON,
		"Expect ';' after variable declaration.",
	); err != nil {
		return stmt.Var[any]{}, err
	}
	return stmt.Var[any]{
		Name:        name,
		Initializer: initializer,
	}, nil
}

func (p *Parser) expression() (expr.Expr[any], error) {
	expression, err := p.assignment()
	if err != nil {
		return nil, err
	}
	return expression, nil
}

func (p *Parser) assignment() (expr.Expr[any], error) {
	expression, err := p.or()
	if err != nil {
		return nil, err
	}

	if p.match(token.EQUAL) {
		equals := p.previous()
		value, err := p.assignment()
		if err != nil {
			return nil, err
		}
		switch e := expression.(type) {
		case expr.Variable[any]:
			return expr.Assign[any]{
				Name:  e.Name,
				Value: value,
			}, nil
		case expr.Get[any]:
			return expr.Set[any]{
				Object: e.Object,
				Name:   e.Name,
				Value:  value,
			}, nil
		default:
			return expression, p.error(equals, "Invalid assignment target.")
		}
	}
	return expression, nil
}

func (p *Parser) or() (expr.Expr[any], error) {
	expression, err := p.and()
	if err != nil {
		return nil, err
	}

	for p.match(token.OR) {
		operator := p.previous()
		right, err := p.and()
		if err != nil {
			return nil, err
		}
		expression = expr.Logical[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) and() (expr.Expr[any], error) {
	expression, err := p.equality()
	if err != nil {
		return nil, err
	}

	for p.match(token.AND) {
		operator := p.previous()
		right, err := p.equality()
		if err != nil {
			return nil, err
		}
		expression = expr.Logical[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) equality() (expr.Expr[any], error) {
	expression, err := p.comparison()
	if err != nil {
		return nil, err
	}

	for p.match(token.EQUAL_EQUAL, token.BANG_EQUAL) {
		operator := p.previous()
		right, err := p.comparison()
		if err != nil {
			return nil, err
		}
		expression = expr.Binary[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) comparison() (expr.Expr[any], error) {
	expression, err := p.term()
	if err != nil {
		return nil, err
	}

	for p.match(
		token.GREATER,
		token.GREATER_EQUAL,
		token.LESS,
		token.LESS_EQUAL,
	) {
		operator := p.previous()
		right, err := p.term()
		if err != nil {
			return nil, err
		}
		expression = expr.Binary[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) term() (expr.Expr[any], error) {
	expression, err := p.factor()
	if err != nil {
		return nil, err
	}

	for p.match(
		token.MINUS,
		token.PLUS,
	) {
		operator := p.previous()
		right, err := p.factor()
		if err != nil {
			return nil, err
		}
		expression = expr.Binary[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) factor() (expr.Expr[any], error) {
	expression, err := p.unary()
	if err != nil {
		return nil, err
	}

	for p.match(
		token.SLASH,
		token.STAR,
	) {
		operator := p.previous()
		right, err := p.unary()
		if err != nil {
			return nil, err
		}
		expression = expr.Binary[any]{
			Left:     expression,
			Operator: operator,
			Right:    right,
		}
	}
	return expression, nil
}

func (p *Parser) unary() (expr.Expr[any], error) {
	if p.match(
		token.BANG,
		token.MINUS,
	) {
		operator := p.previous()
		right, err := p.unary()
		if err != nil {
			return nil, err
		}
		return expr.Unary[any]{
			Operator: operator,
			Right:    right,
		}, nil
	}
	return p.call()
}

func (p *Parser) call() (expr.Expr[any], error) {
	expression, err := p.primary()
	if err != nil {
		return nil, err
	}

	for {
		if p.match(token.LEFT_PAREN) {
			expression, err = p.finishCall(expression)
		} else if p.match(token.DOT) {
			name, err := p.consume(
				token.IDENTIFIER,
				"Expect property name after '.'.",
			)
			if err != nil {
				return nil, err
			}
			expression = expr.Get[any]{
				Object: expression,
				Name:   name,
			}
		} else {
			break
		}
	}
	return expression, nil
}

func (p *Parser) finishCall(expression expr.Expr[any]) (expr.Expr[any], error) {
	arguments := make([]expr.Expr[any], 0)

	if p.check(token.RIGHT_PAREN) {
		argument, err := p.expression()
		if err != nil {
			return nil, err
		}
		arguments = append(arguments, argument)
		for p.match(token.COMMA) {
			if len(arguments) >= 255 {
				return nil, p.error(
					p.peek(),
					"Can't have more than 255 arguments.",
				)
			}
			argument, err := p.expression()
			if err != nil {
				return nil, err
			}
			arguments = append(arguments, argument)
		}
	}

	paren, err := p.consume(token.RIGHT_PAREN, "Expect ')' after arguments.")
	if err != nil {
		return nil, err
	}

	return expr.Call[any]{
		Callee:    expression,
		Paren:     paren,
		Arguments: arguments,
	}, nil
}

func (p *Parser) primary() (expr.Expr[any], error) {
	if p.match(token.TRUE) {
		return expr.Literal[any]{Value: true}, nil
	}
	if p.match(token.FALSE) {
		return expr.Literal[any]{Value: false}, nil
	}
	if p.match(token.NIL) {
		return expr.Literal[any]{Value: nil}, nil
	}
	if p.match(token.IDENTIFIER) {
		return expr.Variable[any]{Name: p.previous()}, nil
	}
	if p.match(token.NUMBER, token.STRING) {
		return expr.Literal[any]{Value: p.previous().Literal}, nil
	}
	if p.match(token.THIS) {
		return expr.This[any]{Keyword: p.previous()}, nil
	}
	if p.match(token.LEFT_PAREN) {
		expression, err := p.expression()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume(
			token.RIGHT_PAREN,
			"Expect ')' after expression.",
		); err != nil {
			return nil, err
		}
		return expr.Grouping[any]{Expression: expression}, nil
	}
	if p.match(token.SUPER) {
		keyword := p.previous()
		if _, err := p.consume(
			token.DOT,
			"Expect '.' after 'super'.",
		); err != nil {
			return nil, err
		}
		method, err := p.consume(token.IDENTIFIER, "Expect superclass method name.")
		if err != nil {
			return nil, err
		}
		return expr.Super[any]{
			Keyword: keyword,
			Method:  method,
		}, nil
	}
	return nil, p.error(p.peek(), "Expect expression.")
}

func (p *Parser) classDeclaration() (stmt.Class[any], error) {
	name, err := p.consume(token.IDENTIFIER, "Expect class name.")
	if err != nil {
		return stmt.Class[any]{}, err
	}
	var superclass expr.Variable[any]
	if p.match(token.LESS) {
		if _, err := p.consume(
			token.IDENTIFIER,
			"Expect superclass name.",
		); err != nil {
			return stmt.Class[any]{}, err
		}
		superclass = expr.Variable[any]{
			Name: p.previous(),
		}
	}
	if _, err := p.consume(
		token.LEFT_BRACE,
		"Expect '{' before class body.",
	); err != nil {
		return stmt.Class[any]{}, err
	}
	methods := make([]stmt.Function[any], 0)
	for p.check(token.RIGHT_BRACE) && !p.isAtEnd() {
		function, err := p.function("method")
		if err != nil {
			return stmt.Class[any]{}, nil
		}
		methods = append(methods, function)
	}
	if _, err := p.consume(
		token.RIGHT_BRACE,
		"Expect '}' after class body.",
	); err != nil {
		return stmt.Class[any]{}, err
	}
	return stmt.Class[any]{
		Name:       name,
		Superclass: superclass,
		Methods:    methods,
	}, nil
}

func (p *Parser) function(kind string) (stmt.Function[any], error) {
	name, err := p.consume(token.IDENTIFIER, fmt.Sprintf("Expect %s name.", kind))
	if err != nil {
		return stmt.Function[any]{}, err
	}
	if _, err := p.consume(
		token.LEFT_PAREN,
		fmt.Sprintf("Expect '(' after %s name.", kind),
	); err != nil {
		return stmt.Function[any]{}, err
	}
	arguments := make([]token.Token, 0)
	if !p.check(token.RIGHT_PAREN) {
		identifer, err := p.consume(token.IDENTIFIER, "Expect parameter name.")
		if err != nil {
			return stmt.Function[any]{}, err
		}
		arguments = append(arguments, identifer)
		for p.match(token.COMMA) {
			if len(arguments) >= 255 {
				p.reportError(p.peek(), "Can't have more than 255 parameters.")
			}
			identifer, err := p.consume(token.IDENTIFIER, "Expect parameter name.")
			if err != nil {
				return stmt.Function[any]{}, err
			}
			arguments = append(arguments, identifer)
		}
	}
	if _, err := p.consume(
		token.RIGHT_PAREN,
		"Expect ')' after parameters.",
	); err != nil {
		return stmt.Function[any]{}, err
	}
	if _, err := p.consume(
		token.LEFT_BRACE,
		fmt.Sprintf("Expect '{' before %s body.", kind),
	); err != nil {
		return stmt.Function[any]{}, err
	}
	body, err := p.block()
	if err != nil {
		return stmt.Function[any]{}, err
	}
	return stmt.Function[any]{
		Name:   name,
		Params: arguments,
		Body:   body,
	}, nil
}

func (p *Parser) block() ([]stmt.Stmt[any], error) {
	statements := make([]stmt.Stmt[any], 0)

	for !p.check(token.RIGHT_BRACE) && !p.isAtEnd() {
		statements = append(statements, p.declaration())
	}
	if _, err := p.consume(
		token.RIGHT_BRACE,
		"Expect '}' after block.",
	); err != nil {
		return nil, err
	}
	return statements, nil
}

func (p *Parser) isAtEnd() bool {
	return p.tokens[p.current].TokenType == token.EOF
}

func (p *Parser) match(types ...token.TokenType) bool {
	for _, t := range types {
		if p.check(t) {
			p.advance()
			return true
		}
	}
	return false
}

func (p *Parser) check(expected token.TokenType) bool {
	if p.isAtEnd() {
		return false
	}
	return p.peek().TokenType == expected
}

func (p *Parser) advance() token.Token {
	if !p.isAtEnd() {
		p.current += 1
	}
	return p.previous()
}

func (p *Parser) previous() token.Token {
	return p.tokens[p.current-1]
}

func (p *Parser) consume(expected token.TokenType, message string) (token.Token, error) {
	if p.check(expected) {
		return p.advance(), nil
	}
	return token.Token{}, p.error(p.peek(), message)
}

func (p *Parser) error(token token.Token, message string) error {
	p.reportError(token, message)
	return &ParseError{}
}

func (p *Parser) peek() token.Token {
	return p.tokens[p.current]
}

func (p *Parser) synchronize() {
	for p.isAtEnd() {
		if p.previous().TokenType == token.SEMICOLON {
			return
		}
		switch p.peek().TokenType {
		case token.CLASS,
			token.FUN,
			token.VAR,
			token.FOR,
			token.IF,
			token.WHILE,
			token.PRINT,
			token.RETURN:
			return
		default:
			p.advance()
		}
	}
}
